app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,241 @@
+    end
+    parts.each_with_index do |part, index|
+      path = new_url.split('/').slice(0, index + 1).join("/")
+      @breadcrumbs << view_context.link_to(part, "/file_manager/#{path}")
+    end
+    absolute_path = check_path_exist(path)
+    if (File.directory?(absolute_path) && check_instructor(absolute_path)) || path == ""
+      populate_directory(absolute_path, new_url)
+      render 'file_manager/index'
+    elsif File.file?(absolute_path) && check_instructor(absolute_path)
+      if File.size(absolute_path) > 1_000_000 || params[:download]
+        send_file absolute_path
+      else
+        mime_type = MimeMagic.by_path(absolute_path).type
+        if mime_type.split('/').first == 'text'
+          @path = path
+          @file = absolute_path.read
+          render :file, formats: :html
+        else
+          send_file(absolute_path,
+                    filename: File.basename(absolute_path),
+                    disposition: 'attachment')
+        end
+      end
+    end
+  end
+
+  def upload
+    upload_file(params[:path].nil? ? "" : params[:path])
+  end
+
+  def delete
+    absolute_path = check_path_exist(params[:path])
+    return unless check_instructor(absolute_path)
+
+    parent = absolute_path.parent
+    raise "Unable to delete courses in the root directory." if parent == BASE_DIRECTORY
+
+    FileUtils.rm_rf(absolute_path)
+  end
+
+  def rename
+    absolute_path = check_path_exist(params[:relative_path])
+    if check_instructor(absolute_path)
+      parent = absolute_path.parent
+      if parent == BASE_DIRECTORY
+        flash[:error] = "Unable to rename courses in the root directory."
+      else
+        dir_name = File.dirname(params[:relative_path])
+
+        if params[:new_name].nil? || params[:new_name].empty?
+          raise ArgumentError, "New name not provided"
+        end
+
+        unless params[:new_name].match(/\A[a-zA-Z0-9_\-.\s]+\Z/)
+          raise ArgumentError, "Invalid characters. Only letters,
+        numbers, underscores, and hyphens are allowed."
+        end
+
+        new_path = safe_expand_path("#{dir_name}/#{params[:new_name]}")
+        parent = new_path.split[0..-2].join('/')
+
+        raise ArgumentError, "A file with that name already exists" if File.exist?(new_path)
+
+        FileUtils.mkdir_p(parent)
+        FileUtils.mv(absolute_path, new_path)
+        flash[:success] = "Successfully renamed file to #{params[:new_name]}"
+      end
+    end
+  rescue ArgumentError => e
+    flash[:error] = e.message
+  end
+
+  def download_tar
+    path = params[:path]&.split("/")&.drop(2)&.join("/")
+    path = CGI.unescape(path)
+    absolute_path = check_path_exist(path)
+    return unless check_instructor(absolute_path)
+
+    if File.directory?(absolute_path)
+      tar_stream = StringIO.new("")
+      Gem::Package::TarWriter.new(tar_stream) do |tar|
+        Dir[File.join(absolute_path, '**', '**')].each do |file|
```

Comment:
`File.join` and `Regexp.escape` (below) both expect strings but absolute_path is a `Pathname`. It would be best to explicitly convert to a string using `.to_s`

Reply:
Changed both of these